🌱  Refactor MHC tests

wfernandes opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
This PR refactors some flaky MHC tests. It is moving us away from Ginkgo to vanilla Go test with envtest.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3539 

**NOTES**
- There are currently three tests that have been marked with `t.Skip`. As of now, they seem to be the most flaky (on my machine). Also there is an opportunity to refactor them using conditions. See #3026 for more info. The refactor of these tests can be done in separate PRs if not part of the work for #3026.
- There are still Ginkgo tests in the `controllers` package. We should refactor those tests as well, so that we can have more consistent tests. This can be done in subsequent PRs. IMO these are great `good-first-issue`s and helps make our test codebase more consistent!

wfernandes commented on controllers/suite_test.go:
```diff
@@ -45,14 +47,93 @@ const (
-	ctx               = context.Background()
+	testEnv  *helpers.TestEnvironment
+	ctx      = context.Background()
+	testEnv1 *helpers.TestEnvironment
```
Since there are other ginkgo/envtest tests in this package, I've created a second testEnv for now to isolate the flakes happening in the MHC tests.

- Can we even have two envtest environments running together? 
- Should we have a separate PR to convert the other ginkgo/envtest style tests in this package to use this vanilla go test pattern?

wfernandes commented on controllers/suite_test.go:
```diff
@@ -45,14 +47,93 @@ const (
+	}).SetupWithManager(testEnv1.Manager, controller.Options{MaxConcurrentReconciles: 1}); err != nil {
+		panic(fmt.Sprintf("Failed to start ClusterCacheReconciler: %v", err))
+	}
+	// TODO: (wfernandes) Do we need these reconcilers for the MHC tests?
```
I believe these reconcilers can be added if we convert the remaining ginkgo/envtest tests to use the vanilla go test pattern. I've commented these out so that we can focus on the bare minimum for what's needed for the MHC tests.

wfernandes replied:
The way some of these tests are written, I don't see how these are "unit" tests anymore. They definitely feel like integration tests since they rely on other reconcilers to set state on certain objects like `Machine` and `MachineSet` in order for the MHC reconciler to work.

wfernandes commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -111,841 +110,968 @@ func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, st
-						machine.Status.LastUpdated = &now
-						Expect(testEnv.Status().Patch(ctx, &machine, machinePatch)).To(Succeed())
-					}
+func TestMachineHealthCheck_Reconcile(t *testing.T) {
```
This is a big diff. But essentially, I'm pulling out each `Specify` block into a separate `t.Run` subtest and isolating all the objects it requires so as to minimize shared state.

wfernandes reviewed (commented):
Just left some comments to clarify some things.

wfernandes commented:
/label needs-ok-to-test

wfernandes commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -111,841 +110,1004 @@ func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, st
-			mhc.Spec.ClusterName = cluster.Name
-			Expect(testEnv.Create(ctx, mhc)).To(Succeed())
+	t.Run("when a machine is paused", func(t *testing.T) {
+		t.Skip("skipping until made stable")
```
There are a few skipped tests. These tests are very unstable. Will review them after some other conditions related work gets merged in.

wfernandes commented on controllers/machinehealthcheck_targets.go:
```diff
@@ -228,7 +228,11 @@ func (r *MachineHealthCheckReconciler) getNodeFromMachine(clusterClient client.R
 	}
 	err := clusterClient.Get(context.TODO(), nodeKey, node)
-	return node, err
+	// if it cannot find a node, send a nil node back...
```
This didn't seem correct.
`getNodeFromMachine` should return a nil node if it cannot find any node because that nil check happens in the `needRemediation` method for the healthCheckTarget.

ncdc replied:
cc @JoelSpeed

JoelSpeed replied:
I think was a hangover from an original version where if the node wasn't found, we would see a non-nil node with no UID which would suggest the node no longer exists, this was replaced by `nodeMissing` as a field on the target struct.

I think therefore, that this change is fine :)

wfernandes replied:
Thanks for the confirmation @JoelSpeed!

wfernandes commented:
This is a seemingly big PR. I'm happy to walkthrough this PR if it helps reviewers.

wfernandes commented on controllers/remote/suite_test.go:
```diff
@@ -41,11 +41,11 @@ var (
 )
 
-func TestAPIs(t *testing.T) {
+func TestGinkgoSuite(t *testing.T) {
```
Not sure where `TestAPIs` came from but I thought this was more clear as to what is happening.

vincepri replied:
Kubebuilder used to (probably still does) generate it like this

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+
+func fakeNodesMachines(
+	g *WithT,
+	n int,
+	healthy bool,
+	assignNodeRef bool,
+	mhc *clusterv1.MachineHealthCheck,
+	c *clusterv1.Cluster,
```
What about making an input struct assigning more expressive names to vars, see we will improve test readability. E.g.
healthy --> forceNodeUnhealthy
assignNodeRef --> skipNodeCreation (inverted, assuming default behaviour for setupMachinesAndFakeNodes is to create)

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	for i := 0; i < n; i++ {
+		machine := newMachine(mhc, c)
+		infraMachine, providerID := newInfraMachine(&machine)
+		node := newNode()
+		if !healthy {
+			setNodeUnhealthy(&node)
+		}
```
Move down into the if?

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+		}
+		// TODO: Add comment on what's happening here. I think we are making
+		// the InfraMachine look like its healthy and ready.
+		infraMachinePatch := client.MergeFrom(infraMachine.DeepCopy())
+		g.Expect(unstructured.SetNestedField(infraMachine.Object, true, "status", "ready")).To(Succeed())
+		g.Expect(testEnv.Status().Patch(ctx, infraMachine, infraMachinePatch)).To(Succeed())
```
Might be I'm missing something, but I find strange to create objects and then patch them (here and down in the func).
Should we instead create the objects, set fields according to the flags, and then create at the end when the objects are fully configured?

wfernandes replied:
TBH I didn't refactor this func and left the original behavior because I was unsure as to why we were doing some of these patches. I had added a bunch of comments asking why we were doing these things as well. 😄 

I can try refactoring this function to see if it can be improved upon further.

fabriziopandini replied:
ok, good catch.
AFAIK no, you can't set the status during create (status is to be set by the controllers).

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	}
+}
+
+func newInfraMachine(machine *clusterv1.Machine) (*unstructured.Unstructured, string) {
```
nit: given that `newInfraMachine` is always called after `newMachine`, should we merge the two function and get `newMachine` to return both machine and infra machine?

wfernandes replied:
I left this as is because then we also have the providerID to consider.

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	}
+
+	cleanup := func() {
+		fmt.Println("Cleaning up nodes and machines")
```
It seems we are not cleaning up infra machines

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+			g.Expect(testEnv.Create(ctx, &node)).To(Succeed())
+			fmt.Printf("node created: %s\n", node.GetName())
+
+			nodePatch := client.MergeFrom(node.DeepCopy())
+			node.Status = nodeStatus
+			g.Expect(testEnv.Status().Patch(ctx, &node, nodePatch)).To(Succeed())
```
why are we setting again the node status using a patch?

wfernandes replied:
Yeah...I was asking myself this question as well 😄 . I didn't know if there was something interesting happening in the controllers that this needed to be set specifically for the MHC controller.

wfernandes replied:
After looking at it some more I noticed that we are using the status client to patch the object. I'm wondering if we cannot create an object with Status fields set and we need the status client specifically to patch them in. This is just a guess.

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+			nodes = append(nodes, &node)
+
+			machinePatch := client.MergeFrom(machine.DeepCopy())
+			machine.Status = machineStatus
```
why are we setting again machine status?

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+			// adding a second to ensure there is a difference from the
+			// original time so that the patch works.
+			lastUp := metav1.NewTime(machine.Status.LastUpdated.Add(time.Second))
+			machine.Status.LastUpdated = &lastUp
```
why are we altering status.LastUpdated?

wfernandes replied:
LastUpdated needs to be set on the machine so that some logic in the MHC controller (_I believe in `healthCheckTarget/needsRemediation()`_) can be tested. I tried setting this field on the machine object when it was created in the beginning but for some reason the cache wasn't getting updated. It only seemed to work when we patched it like this. Maybe I'm missing something or we just need to wait a little longer until the cache gets updated.

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	}, providerID
+}
+
+func fakeNodesMachines(
```
second thought on this:
this func implements  a different pattern than the one in newXX or in setup, because in this case we are returning some of the objects and a cleanup func, while in both the other cases we are not returning a cleanup func.
What about dropping the cleanup parameter ad return the infrastructureMachines instead

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
 	}
 }
+
+func setup(g *WithT) (*clusterv1.Cluster, *clusterv1.MachineHealthCheck) {
```
```suggestion
func createNamaspaceClusterAndMHC(g *WithT) (*clusterv1.Cluster, *clusterv1.MachineHealthCheck) {
```
I find kind of confusing this func creating ns and cluster, while returning the MHC without creating int.what about adding an an input parameter providing a func to mutate the mhc, so we can create it inside this func?

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,859 @@ import (
+		ctx := context.TODO()
+		cluster, mhc := setup(g)
+
+		mhc.Spec.ClusterName = cluster.Name
```
I see this line in all tests, what about moving into setup?

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,859 @@ import (
-						Expect(testEnv.Status().Patch(ctx, &machine, machinePatch)).To(Succeed())
-					}
+func TestMachineHealthCheck_Reconcile(t *testing.T) {
+	t.Run("it should ensure the correct cluster-name label with no existing labels exist", func(t *testing.T) {
```
```suggestion
	t.Run("it should ensure the correct cluster-name label when no existing labels exist", func(t *testing.T) {
```

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1382,232 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
 }
+
+func setup(g *WithT) (*clusterv1.Cluster, *clusterv1.MachineHealthCheck) {
+	ns, err := testEnv.CreateNamespace(ctx, "test-ns")
```
```suggestion
	ns, err := testEnv.CreateNamespace(ctx, "test-mhc")
```

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,859 @@ import (
+
+		mhc.Spec.ClusterName = cluster.Name
+		mhc.Labels = map[string]string{}
+		mhc.SetGenerateName("correct-cluster-name-label-with-no-existing-labels-")
```
being in a separated ns, might be that having a specific generated name for each test does not add much values. Should we drop?

fabriziopandini reviewed (commented):
@wfernandes thanks for this PR! 
I made a first pass focusing on stream-lining how do we create test objects, and leaving some comments trying to make it explicit what is generated for each test, and to make consistent how do we create/cleanup.
I will make another pass on test content once we finalize ^^

wfernandes commented on controllers/suite_test.go:
```diff
@@ -70,61 +57,85 @@ var _ = BeforeSuite(func(done Done) {
+
+// TestGinkgoSuite will run the ginkgo tests.
+// This will run with the testEnv setup and teardown in TestMain.
+func TestGinkgoSuite(t *testing.T) {
```
Yes. This runs the Ginkgo tests in this package that haven't yet been migrated to go test. I'm waiting for this PR to merge in before I incrementally start refactoring the ginkgo tests to the new style.

I found the original `TestAPIs` to be confusing so I just made it clear that this was the go test that kicks off the ginkgo suite.

wfernandes commented:
@fabriziopandini Thanks for the feedback!

I did some of the refactors to clean up the code and make it more readable. I left some of the stuff back because I can see better opportunities of refactoring when we convert the other ginkgo tests to go testing as well. For example, the code in `suite_util_test.go`. 

Also maybe we can do some of the other refactors in separate smaller PRs since this one is already massive! WDYT?

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1383,8 +1371,10 @@ func ownerReferenceForCluster(ctx context.Context, g *WithT, c *clusterv1.Cluste
-	ns, err := testEnv.CreateNamespace(ctx, "test-ns")
+// setup creates a namespace in the test environment. It then creates a
+// Cluster and KubeconfigSecret for that cluster in said namespace.
+func setup(g *WithT) *clusterv1.Cluster {
```
```suggestion
func createNamespaceAndCluster(g *WithT) *clusterv1.Cluster {
```

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1582,25 +1551,35 @@ func fakeNodesMachines(
 }
 
-func newTestMachineHealthCheck(name, namespace, cluster string, labels map[string]string) *clusterv1.MachineHealthCheck {
+func newMHCWithLabels(name, namespace, cluster string, labels map[string]string) *clusterv1.MachineHealthCheck {
```
```suggestion
func newMachineHealthCheckWithLabels(name, namespace, cluster string, labels map[string]string) *clusterv1.MachineHealthCheck {
```
given that in all the other places we are using the full name

fabriziopandini commented:
> Also maybe we can do some of the other refactors in separate smaller PRs since this one is already massive! WDYT?

@wfernandes ok, I leave to you the option to draw the line; I'm just focusing on this area because in the previous version it was really difficult to understand which object where to create during test setup, and as a consequence it was difficult to debug and understand tests.

fabriziopandini commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
-	}
-	return node
-}
+func TestMachineHealthCheck_Reconcile(t *testing.T) {
```
nit: might be worth to consider to break down tests in more than one Test func

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
-					HaveLen(1),
-				))
-			})
+	t.Run("it marks unhealthy machines for remediation when all Machines are healthy", func(t *testing.T) {
```
Maybe rename this "it doesn't mark anything unhealthy when all Machines are healthy"?

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
+			g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+		}(cluster, mhc, infraTmpl, machineSet)
+
+		// Pause the MachineSet reconciler. WHY?? I SEE WHAT IS HAPPENING BUT
```
This is to delay the deletion of the Machine, which the MachineSet controller does when it sees a Machine marked unhealthy by MHC.

wfernandes replied:
Whoa! Thought I removed all of my ranting comments. Thanks for catching this. 😬

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -1460,3 +1370,235 @@ func ownerReferenceForCluster(ctx context.Context, c *clusterv1.Cluster) metav1.
+	g.Eventually(func() error {
+		var cl clusterv1.Cluster
+		return testEnv.Get(ctx, util.ObjectKey(cluster), &cl)
+	}, timeout, time.Second).Should(Succeed())
```
Can we poll more frequently than once per second?

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
+				return nil
+			}
+			return mhc.GetOwnerReferences()
+		}, timeout, time.Second).Should(And(
```
Increase poll frequency?

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
+				return nil
+			}
+			return mhc.GetOwnerReferences()
+		}, timeout, time.Second).Should(And(
```
Increase poll frequency?

ncdc commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,847 @@ import (
+				return nil
+			}
+			return &mhc.Status
+		}, timeout, time.Second).Should(MatchMachineHealthCheckStatus(&clusterv1.MachineHealthCheckStatus{
```
Increase poll frequency? (Will stop commenting, but same comment applies to all one-second frequencies).

wfernandes commented:
@ncdc Thanks for the review! I reduced the polling times to `100ms` and updated some of the comments.
As per our discussion, I also changed the function signature of `createMachinesWithNodes` to include optional function arguments to make it more readable.

wfernandes commented on controllers/machinehealthcheck_controller_test.go:
```diff
@@ -46,906 +45,911 @@ import (
-			}).Should(Equal(1))
-
-			// Unpause the MachineSet reconciler.
-			machineSetPatch = client.MergeFrom(machineSet.DeepCopy())
-			delete(machineSet.Annotations, clusterv1.PausedAnnotation)
-			Expect(testEnv.Patch(ctx, machineSet, machineSetPatch)).To(Succeed())
-
-			// Make sure the Machine gets deleted.
-			Eventually(func() bool {
-				machine := unhealthyMachine.DeepCopy()
-				err := testEnv.Get(ctx, util.ObjectKey(unhealthyMachine), machine)
-				return apierrors.IsNotFound(err) || !machine.DeletionTimestamp.IsZero()
+			return len(machines.Items)
+		}, timeout, 100*time.Millisecond).Should(Equal(1))
```
So I just noticed this as I was reviewing another PR. 
I'm guessing either the comment for this block is incorrect or the logic inside the eventually is incorrect and needs to match something similar to line 684.

@ncdc or anyone confirm this?

ncdc replied:
I believe the comment on 779 should be reworded to say we're waiting to make sure the MachineSet controller has created the required Machine from the MachineSet.

I think you can also copy the block from L684 to just before the block at L826

ncdc commented:
/lgtm

vincepri reviewed (approved):
/approve
/milestone v0.3.10
